### ISSUE TITLE
False positive with generic default argument.

### ISSUE THREAD

**Issue by parched** [microsoft/pyright#4801]
While upgrading pyright I noticed this new error which I believe should not be an error.

**Screenshots or Code**
```python

_T = TypeVar("_T")
_D = TypeVar("_D")


def _get(
    value: Mapping[str, object], name: str, type_: type[_T], default: _D = None
) -> _T | _D:
    """Return a value from the mapping and check the type or default if it doesn't exist"""
    val = value.get(name)
    if val is None:
        return default

    if not isinstance(val, type_):
        raise TypeError(
            f"{name} should be type {type_.__name__} not type {type(val).__name__}"
        )

    return val

```
![image](https://user-images.githubusercontent.com/5975405/226152141-0376f4a5-b813-4635-913d-0fdc35a6e926.png)

**VS Code extension or command-line**
Commandline. It starting happening in 1.1.282

**Comment by erictraut** [comment]
This behavior is by design. When assigning a default argument to a parameter whose type is a "naked" TypeVar, the effective type is modified to be a union of the TypeVar and the type of the argument type. Mypy doesn't allow this case (where a parameter with a "naked" TypeVar has a default argument value) because it's generally unsafe. Pyright allows it but with this restriction.

**Comment by parched** [comment]
How should I annotate this function then? Do I need to use `overload`?

**Comment by erictraut** [comment]
You could use an overload, but that wouldn't help with the type checking of the implementation. You'll probably need to use a cast.
